README
=GCal4Ruby
==Introduction
GCal4Ruby is a full featured wrapper for the google calendar API.  GCal4Ruby implements
all of the functionality available through the Google Calnedar API, including permissions,
attendees, reminders and event recurrence.  

==Author and Contact Information
This branch was made by Matt Gornick (http://github.com/mgornick) to implement Google AuthSub into the original 
GCal4Ruby that was created and is maintained by {Mike Reich}[mailto:mike@seabourneconsulting.com] 
and is licenses under the GPL v2.  Feel free to use and update, but be sure to contribute your
code back to the project and attribute as required by the license.
===Website
http://rubyforge.org/projects/gcal4ruby/

===Example Application
http://github.com/mgornick/studentplanr

==Description
GCal4Ruby has three major components: the service, calendar and event objects.  Each service
has many calendars, which in turn have many events.  Each service is the representation of a
google account, and thus must be successfully authenticated using valid Google Calendar
account credentials.  

==Examples
Below are some common usage examples.  For more examples, check the documentation.
===Service
1. Authenticate
    service = Service.new
    service.authenticate("user@gmail.com", "password")

2. Authenticate with Google AuthSub
		# your authentication controller (example is index action in the home controller):
			scope = 'http://www.google.com/calendar/feeds/'
	    next_url = 'http://127.0.0.1:3000/home/authenticated'
	    secure = false  # set secure = true for signed AuthSub requests
	    session = true
	    @authsub_link = GData::Auth::AuthSub.get_url(next_url, scope, secure, session)

		# in your next_url controller as defined by your AuthSub link (in example would be the authenticated action)
			@token = params[:token]
  		client = GData::Client::Calendar.new
  		client.authsub_token = params[:token] # extract the single-use token from the URL query params
  		session[:token] = client.auth_handler.upgrade()
  		client.authsub_token = session[:token] if session[:token]
  		
  		service= GCal4Ruby::Service.new
  		service.authsub_authenticate(session[:token], "user@gmail.com")

3. Get Calendar List
    calendars = service.calendars

===Calendar
All usages assume a successfully authenticated Service.
1. Create a new Calendar
    cal = Calendar.new(service)

2. Find an existing Calendar
    cal = Calendar.find(service, "New Calendar", :first)

3. Find all calendars containing the search term
    cal = Calendar.find(service, "Soccer Team")

4. Find a calendar by ID
    cal = Calendar.find(service, id, :first)
===Event
All usages assume a successfully authenticated Service and valid Calendar.
1. Create a new Event
    event = Event.new(calendar)
    event.title = "Soccer Game"
    event.start = Time.parse("12-06-2009 at 12:30 PM")
    event.end = Time.parse("12-06-2009 at 1:30 PM")
    event.where = "Merry Playfields"
    event.save

2. Find an existing Event
    event = Event.find(cal, "Soccer Game", {:scope => :first})

3. Find all events containing the search term
    event = Event.find(cal, "Soccer Game")

4. Create a recurring event for every saturday
    event = Event.new(calendar)
    event.title = "Baseball Game"
    event.where = "Municipal Stadium"
    event.recurrence = Recurrence.new
    event.recurrence.start = Time.parse("13-06-2009 at 4:30 PM")
    event.recurrence.end = Time.parse("13-06-2009 at 6:30 PM")
    event.recurrence.frequency = {"weekly" => ["SA"]}
    event.save 

5. Create an event with a 15 minute email reminder
    event = Event.new(calendar)
    event.title = "Dinner with Kate"
    event.start = Time.parse("20-06-2009 at 5 pm")
    event.end = Time.parse("20-06-209 at 8 pm")
    event.where = "Luigi's"
    event.reminder = {:minutes => 15, :method => 'email'}
    event.save

6. Create an event with attendees
    event = Event.new(calendar)
    event.title = "Dinner with Kate"
    event.start = Time.parse("20-06-2009 at 5 pm")
    event.end = Time.parse("20-06-209 at 8 pm")
    event.attendees => {:name => "Kate", :email => "kate@gmail.com"}
    event.save


CHANGELOG
#=CHANGELOG
#==version 0.3.1_AuthSub
#* Git repo by mgornick to implement Authsub authentication into the GCal4Ruby gem
#==version 0.3.1
#* Fixed Event.find method to work with secondary calendars and google apps accounts.  Fix provided by groesser3.
#* Added max results to Calendar.find.
#==version 0.3.0
#* Rewrote Event.find to improve performance significantly.
#* Added improvements to event recurrence handling, including loading existing recurrences, changing recurring events to non recurring and vice versa.
#* Added support for initialization attributes to Event, Calendar, Service and Recurrence.
#* Fixed query string typo in Event.find.  Fix provided by nat.lownes.
#==version 0.2.11
#* Added support for GML elements in calendar events.  Fix provided by nat.lownes.
#* Fixed event status bug where cancelled events were marked confirmed.  Fix provided by rifraf.
#==version 0.2.10
#* Fixed library support for Google Hosted Apps accounts running in forced SSL mode.
#==version 0.2.9
#* Fixed small SSL redirect bug due to variable misnaming in base.rb.  Fix provided by JohnMetta
#==version 0.2.8
#* Merged changes from edruder and h13ronim from the unofficial github repo - http://github.com/h13ronim/gcal4ruby/commits/master
#==version 0.2.7
#* Added fix for finding events in calendars that have more than 25 entries
#==version 0.2.6
#* Added fix for updated google calendar XML per http://cookingandcoding.wordpress.com/2009/06/08/new-ruby-google-calendar-api-gem-gcal4ruby/#comment-183
#==version 0.2.5
#* Added calendar color support to to_iframe methods in Calendar and Service.
#==version 0.2.4
#* Fixed bug with ACL check in Calendar#load
#==version 0.2.3
#* Implemented to_iframe method for calendars and services to output embeddable iframe text.
#* Added switch to turn off ACL check for public calendars.  Can increase effeciency if turned off.
#==version 0.2.2
#* Fixed URL encoding problem in Event.find method.
#* cleaned up Event.find method to allow for finding events by id
#* updated Calendar.find method to add params hash
#* added 'published', 'updated', and 'edited' attributes
#==version 0.2.1
#* fixed Event.find calendar specification
#==version 0.2.0
#* Fixed redirect URI query params problem
#* Updated syntax for finding events to include most google api parameters,  Non-backwards compatible.
#==version 0.1.4
#* Added additional search criteria for Event.find
#==version 0.1.3
#* Added support for authenticating with Google Hosted Apps accounts
#* Added flag to indicate whether a calendar is editable
#* Added handling to gracefully throw error when trying to create event on a non-editable (shared) calendar
#==version 0.1.2
#* Fixed to_xml dump problem with hidden and selected attributes
#==version 0.1.1
#* Added all_day indicator to event to indicate an all day event
#==version 0.1.0
#* Initial Version

test/unit.rb
#!/usr/bin/ruby

require 'rubygems'
require 'gcal4ruby'
include GCal4Ruby

@service = Service.new
@username = nil
@password = nil

def tester
  if ARGV.include?("-d")
      @service.debug = true
  end
  ARGV.each do |ar|
    if ar.match("username=")
      @username = ar.gsub("username=", "")
    end
    if ar.match("password=")
      @password = ar.gsub("password=", "")
    end
  end
  service_test
  calendar_test
  event_test
  event_recurrence_test
end

def service_test
  puts "---Starting Service Test---"
  puts "1. Authenticate"
  if @service.authenticate(@username, @password)
    successful
  else
    failed
  end
  
  puts "2. Calendar List"
  cals = @service.calendars
  if cals
    successful "Calendars for this Account:"
    cals.each do |cal|
      puts cal.title
    end
  else
    failed
  end
end

def calendar_test
  puts "---Starting Calendar Test---"
  
  puts "1. Create Calendar"
  cal = Calendar.new(@service)
  cal.title = "test calendar"+Time.now.to_s
  puts "Calender exists = "+cal.exists?.to_s
  if cal.save
    successful cal.to_xml
  else
    failed
  end
  
  puts "2. Edit Calendar"
  cal.title = "renamed title"
  if cal.save
    successful cal.to_xml
  else
    puts "Test 2 Failed"
  end
  
  puts "3. Find Calendar by ID"
  c = Calendar.find(@service, cal.id)
  if c.title == cal.title
    successful
  else
    failed "#{c.title} not equal to #{cal.title}"
  end
  
  puts "4. Delete Calendar"
  if cal.delete and not cal.title
    successful
  else
    failed
  end
end

def event_test
  puts "---Starting Event Test---"
  
  puts "1. Create Event"
  event = Event.new(@service.calendars[0])
  event.title = "Test Event"
  event.content = "Test event content"
  event.start = Time.now+1800
  event.end = Time.now+5400
  if event.save
    successful event.to_xml
  else
    failed
  end
  
  puts "2. Edit Event"
  event.title = "Edited title"
  if event.save
    successful event.to_xml
  else
    failed
  end
  
  puts "3. Reload Event"
  if event.reload
    successful
  end
  
  puts "4. Find Event by id"
  e = Event.find(@service.calendars[0], event.id)
  if e.title == event.title
    successful
  else
    failed "Found event doesn't match existing event"
  end
  
  puts "5. Delete Event"
  if event.delete
    successful 
  else
    failed
  end
end

def event_recurrence_test
  puts "---Starting Event Recurrence Test---"
  
  @first_start = Time.now
  @first_end = Time.now+3600
  @first_freq = {'weekly' => ['TU']}
  @second_start = Time.now+86000
  @second_end = Time.now+89600
  @second_freq = {'weekly' => ['SA']}
  
  puts "1. Create Recurring Event"
  event = Event.new(@service.calendars[0])
  event.title = "Test Recurring Event"
  event.content = "Test event content"
  event.recurrence = Recurrence.new({:start => @first_start, :end => @first_end, :frequency => @first_freq})
  if event.save 
    successful event.to_xml
  else
    failed("recurrence = "+event.recurrence.to_s)
  end
  
  puts "2. Edit Recurrence"
  event.title = "Edited recurring title"
  event.recurrence = Recurrence.new({:start => @second_start, :end => @second_end, :frequency => @second_freq})
  if event.save 
    successful event.to_xml
  else
    failed
  end
  
  puts "3. Delete Event"
  if event.delete
    successful 
  else
    failed
  end
end

def failed(m = nil)
  puts "Test Failed"
  puts m if m
  exit()
end

def successful(m = nil)
  puts "Test Successful"
  puts m if m
end

tester

lib/gcal4ruby/service.rb
require 'gcal4ruby/base' 
require 'gcal4ruby/calendar'

module GCal4Ruby

#The service class is the main handler for all direct interactions with the 
#Google Calendar API.  A service represents a single user account.  Each user
#account can have multiple calendars, so you'll need to find the calendar you
#want from the service, using the Calendar#find class method.
#=Usage
#
#1. Authenticate
#    service = Service.new
#    service.authenticate("user@gmail.com", "password")
#
#2. Get Calendar List
#    calendars = service.calendars
#

class Service < Base
  #Convenience attribute contains the currently authenticated account name
  attr_accessor :account
      
  # The token returned by the Google servers, used to authorize all subsequent messages
  attr_accessor :auth_token
  
  # Determines whether GCal4Ruby ensures a calendar is public.  Setting this to false can increase speeds by 
  # 50% but can cause errors if you try to do something to a calendar that is not public and you don't have
  # adequate permissions
  attr_accessor :check_public
  
  #added auth type to account for differences in headers between ClientLogin and Authsub
  attr_accessor :auth_type
  
  #Accepts an optional attributes hash for initialization values
  def initialize(attributes = {})
    super()
    attributes.each do |key, value|
      self.send("#{key}=", value)
    end    
    @check_public ||= true
  end

  # The authenticate method passes the username and password to google servers.  
  # If authentication succeeds, returns true, otherwise raises the AuthenticationFailed error.
  def authenticate(username, password)
    ret = nil
    ret = send_post(AUTH_URL, "Email=#{username}&Passwd=#{password}&source=GCal4Ruby&service=cl&accountType=HOSTED_OR_GOOGLE")
    if ret.class == Net::HTTPOK
      @auth_token = ret.read_body.to_a[2].gsub("Auth=", "").strip
      @account = username
      @auth_type = 'ClientLogin'
      return true
    else
      raise AuthenticationFailed
    end
  end
  
  # added authsub authentication.  pass in the upgraded authsub token and the username/email address
  def authsub_authenticate(authsub_token, account)
    @auth_token = authsub_token
    @account = account
    @auth_type = 'AuthSub'
    return true
  end

  #Returns an array of Calendar objects for each calendar associated with 
  #the authenticated account.
  def calendars
    if not @auth_token
       raise NotAuthenticated
    end
    ret = send_get(CALENDAR_LIST_FEED+"?max-results=10000")
    cals = []
    REXML::Document.new(ret.body).root.elements.each("entry"){}.map do |entry|
      entry.attributes["xmlns:gCal"] = "http://schemas.google.com/gCal/2005"
      entry.attributes["xmlns:gd"] = "http://schemas.google.com/g/2005"
      entry.attributes["xmlns:app"] = "http://www.w3.org/2007/app"
      entry.attributes["xmlns"] = "http://www.w3.org/2005/Atom"
      cal = Calendar.new(self)
      cal.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
      cals << cal
    end
    return cals
  end
  
  #Helper function to return a formatted iframe embedded google calendar.  Parameters are:
  #1. *cals*: either an array of calendar ids, or <em>:all</em> for all calendars, or <em>:first</em> for the first (usally default) calendar
  #2. *params*: a hash of parameters that affect the display of the embedded calendar:
  #   height:: the height of the embedded calendar in pixels
  #   width:: the width of the embedded calendar in pixels
  #   title:: the title to display
  #   bgcolor:: the background color.  Limited choices, see google docs for allowable values.
  #   color:: the color of the calendar elements.  Limited choices, see google docs for allowable values.
  #   showTitle:: set to 'false' to hide the title
  #   showDate:: set to 'false' to hide the current date
  #   showNav:: set to 'false to hide the navigation tools
  #   showPrint:: set to 'false' to hide the print icon
  #   showTabs:: set to 'false' to hide the tabs
  #   showCalendars:: set to 'false' to hide the calendars selection drop down
  #   showTimezone:: set to 'false' to hide the timezone selection
  #   border:: the border width in pixels
  #   dates:: a range of dates to display in the format of 'yyyymmdd/yyyymmdd'.  Example: 20090820/20091001
  #   privateKey:: use to display a private calendar.  You can find this key under the calendar settings pane of the Google Calendar website.
  #   colors:: a hash of calendar ids as key and color values as associated hash values.  Example: {'test@gmail.com' => '#7A367A'} 
  def to_iframe(cals, params = {})
    params[:height] ||= "600"
    params[:width] ||= "600"
    params[:title] ||= (self.account ? self.account : '')
    params[:bgcolor] ||= "#FFFFFF"
    params[:color] ||= "#2952A3"
    params[:showTitle] = params[:showTitle] == false ? "showTitle=0" : ''
    params[:showNav] = params[:showNav] == false ? "showNav=0" : ''
    params[:showDate] = params[:showDate] == false ? "showDate=0" : ''
    params[:showPrint] = params[:showPrint] == false ? "showPrint=0" : ''
    params[:showTabs] = params[:showTabs] == false ? "showTabs=0" : ''
    params[:showCalendars] = params[:showCalendars] == false ? "showCalendars=0" : ''
    params[:showTimezone] = params[:showTimezone] == false ? 'showTz=0' : ''
    params[:border] ||= "0"
    output = ''
    puts "params = #{params.inspect}" if self.debug
    params.each do |key, value|
      puts "key = #{key} and value = #{value}" if self.debug
      case key
        when :height then output += "height=#{value}"
        when :width then output += "width=#{value}"
        when :title then output += "title=#{CGI.escape(value)}"
        when :bgcolor then output += "bgcolor=#{CGI.escape(value)}"
        when :color then output += "color=#{CGI.escape(value)}"
        when :showTitle then output += value
        when :showDate then output += value
        when :showNav then output += value
        when :showPrint then output += value
        when :showTabs then output += value
        when :showCalendars then output += value
        when :showTimezone then output += value
        when :viewMode then output += "mode=#{value}"
        when :dates then output += "dates=#{CGI.escape(value)}"
        when :privateKey then output += "pvttk=#{value}"
      end
      output += "&amp;"
    end
    
    puts "output = #{output}" if self.debug
    
    if cals.is_a?(Array)
      for c in cals
        output += "src=#{c}&amp;"
        if params[:colors] and params[:colors][c]
          output += "color=#{CGI.escape(params[:colors][c])}&amp;"
        end
      end
    elsif cals == :all
      cal_list = calendars()
      for c in cal_list
        output += "src=#{c.id}&amp;"
      end
    elsif cals == :first
      cal_list = calendars()
      output += "src=#{cal_list[0].id}&amp;"
    end
        
    "<iframe src='http://www.google.com/calendar/embed?#{output}' style='#{params[:border]} px solid;' width='#{params[:width]}' height='#{params[:height]}' frameborder='#{params[:border]}' scrolling='no'></iframe>"
  end
end

end

lib/gcal4ruby/event.rb
require 'gcal4ruby/recurrence'

module GCal4Ruby
  #The Event Class represents a remote event in calendar. 
  #
  #=Usage
  #All usages assume a successfully authenticated Service and valid Calendar.
  #1. Create a new Event
  #    event = Event.new(calendar)
  #    event.title = "Soccer Game"
  #    event.start = Time.parse("12-06-2009 at 12:30 PM")
  #    event.end = Time.parse("12-06-2009 at 1:30 PM")
  #    event.where = "Merry Playfields"
  #    event.save
  #
  #2. Find an existing Event
  #    event = Event.find(cal, "Soccer Game", {:scope => :first})
  #
  #3. Find all events containing the search term
  #    event = Event.find(cal, "Soccer Game")
  #
  #4. Create a recurring event for every saturday
  #    event = Event.new(calendar)
  #    event.title = "Baseball Game"
  #    event.where = "Municipal Stadium"
  #    event.recurrence = Recurrence.new
  #    event.recurrence.start = Time.parse("13-06-2009 at 4:30 PM")
  #    event.recurrence.end = Time.parse("13-06-2009 at 6:30 PM")
  #    event.recurrence.frequency = {"weekly" => ["SA"]}
  #    event.save 
  #
  #5. Create an event with a 15 minute email reminder
  #    event = Event.new(calendar)
  #    event.title = "Dinner with Kate"
  #    event.start = Time.parse("20-06-2009 at 5 pm")
  #    event.end = Time.parse("20-06-209 at 8 pm")
  #    event.where = "Luigi's"
  #    event.reminder = {:minutes => 15, :method => 'email'}
  #    event.save
  #
  #6. Create an event with attendees
  #    event = Event.new(calendar)
  #    event.title = "Dinner with Kate"
  #    event.start = Time.parse("20-06-2009 at 5 pm")
  #    event.end = Time.parse("20-06-209 at 8 pm")
  #    event.attendees => {:name => "Kate", :email => "kate@gmail.com"}
  #    event.save
  #
  #After an event object has been created or loaded, you can change any of the 
  #attributes like you would any other object.  Be sure to save the event to write changes
  #to the Google Calendar service.
  class Event
    #The event title
    attr_accessor :title
    #The content for the event
    attr_accessor :content
    #The location of the event
    attr_accessor :where
    #A flag for whether the event show as :free or :busy
    attr_accessor :transparency
    #A flag indicating the status of the event.  Values can be :confirmed, :tentative or :cancelled
    attr_accessor :status
    #The unique event ID
    attr_accessor :id
    #Flag indicating whether it is an all day event
    attr_accessor :all_day
    
    @attendees
    
    #The event start time
    attr_reader :start
    #The event end time
    attr_reader :end
    #The reminder settings for the event, returned as a hash
    attr_reader :reminder
    #The date the event was created
    attr_reader :published
    #The date the event was last updated
    attr_reader :updated
    #The date the event was last edited
    attr_reader :edited
    
    #Sets the reminder options for the event.  Parameter must be a hash containing one of 
    #:hours, :minutes and :days, which are simply the number of each before the event start date you'd like to 
    #receive the reminder.  
    #
    #:method can be one of the following:
    #- <b>'alert'</b>: causes an alert to appear when a user is viewing the calendar in a browser
    #- <b>'email'</b>: sends the user an email message
    def reminder=(r)
      @reminder = r
    end
    
    #Returns the current event's Recurrence information
    def recurrence
      @recurrence
    end
    
    #Returns an array of the current attendees
    def attendees
      @attendees
    end
    
    #Accepts an array of email address/name pairs for attendees.  
    #  [{:name => 'Mike Reich', :email => 'mike@seabourneconsulting.com'}]
    #The email address is requried, but the name is optional
    def attendees=(a)
      if a.is_a?(Array)
        @attendees = a
      else
        raise "Attendees must be an Array of email/name hash pairs"
      end
    end
    
    #Sets the event's recurrence information to a Recurrence object.  Returns the recurrence if successful,
    #false otherwise
    def recurrence=(r)
      if r.is_a?(Recurrence) or r.nil?
        r.event = self unless r.nil?
        @recurrence = r
      else
        return false
      end
    end
    
    #Returns a duplicate of the current event as a new Event object
    def copy()
      e = Event.new()
      e.load(to_xml)
      e.calendar = @calendar
      return e
    end
    
    #Sets the start time of the Event.  Must be a Time object or a parsable string representation
    #of a time.
    def start=(str)
      if str.is_a?String
        @start = Time.parse(str)      
      elsif str.is_a?Time
        @start = str
      else
        raise "Start Time must be either Time or String"
      end
    end
    
    #Sets the end time of the Event.  Must be a Time object or a parsable string representation
    #of a time.
    def end=(str)
      if str.is_a?String
        @end = Time.parse(str)      
      elsif str.is_a?Time
        @end = str
      else
        raise "End Time must be either Time or String"
      end
    end
    
    #Deletes the event from the Google Calendar Service.  All values are cleared.
    def delete
        if @exists    
          if @calendar.service.send_delete(@edit_feed, {"If-Match" => @etag})
            @exists = false
            @deleted = true
            @title = nil
            @content = nil
            @id = nil
            @start = nil
            @end = nil
            @transparency = nil
            @status = nil
            @where = nil
            return true
          else
            return false
          end
        else
          return false
        end
    end
    
    #Creates a new Event.  Accepts a valid Calendar object and optional attributes hash.
    def initialize(calendar, attributes = {})
      if not calendar.editable
        raise CalendarNotEditable
      end
      super()
      attributes.each do |key, value|
        self.send("#{key}=", value)
      end
      @xml ||= EVENT_XML
      @calendar ||= calendar
      @transparency ||= "http://schemas.google.com/g/2005#event.opaque"
      @status ||= "http://schemas.google.com/g/2005#event.confirmed"
      @attendees ||= []
      @all_day ||= false
    end
    
    #If the event does not exist on the Google Calendar service, save creates it.  Otherwise
    #updates the existing event data.  Returns true on success, false otherwise.
    def save
      if @deleted
        return false
      end
      if @exists 
        ret = @calendar.service.send_put(@edit_feed, to_xml, {'Content-Type' => 'application/atom+xml', "If-Match" => @etag})
      else
        ret = @calendar.service.send_post(@calendar.event_feed, to_xml, {'Content-Type' => 'application/atom+xml'})
      end
      if !@exists
        if load(ret.read_body)
          return true
        else
          raise EventSaveFailed
        end
      end
      reload
      return true
    end
    
    #Returns an XML representation of the event.
    def to_xml()
      xml = REXML::Document.new(@xml)
      xml.root.elements.each(){}.map do |ele|
        case ele.name
        when 'id'
          ele.text = @id
        when "title"
          ele.text = @title
        when "content"
          ele.text = @content
        when "when"
          if not @recurrence
            ele.attributes["startTime"] = @all_day ? @start.strftime("%Y-%m-%d") : @start.xmlschema
            ele.attributes["endTime"] = @all_day ? @end.strftime("%Y-%m-%d") : @end.xmlschema
            set_reminder(ele)
          else
            if not @reminder
              xml.root.delete_element("/entry/gd:when")
              xml.root.add_element("gd:recurrence").text = @recurrence.to_s
            else
              ele.delete_attribute('startTime')
              ele.delete_attribute('endTime')
              set_reminder(ele)  
            end
          end
        when "eventStatus"
          ele.attributes["value"] = case @status
            when :confirmed
              "http://schemas.google.com/g/2005#event.confirmed"
            when :tentative
              "http://schemas.google.com/g/2005#event.tentative"
            when :cancelled
              "http://schemas.google.com/g/2005#event.canceled"
            else
              "http://schemas.google.com/g/2005#event.confirmed"
          end
        when "transparency"
          ele.attributes["value"] = case @transparency
              when :free
                "http://schemas.google.com/g/2005#event.transparent"
              when :busy
                "http://schemas.google.com/g/2005#event.opaque"
              else
                "http://schemas.google.com/g/2005#event.opaque"
            end
        when "where"
          ele.attributes["valueString"] = @where
        when "recurrence"
          puts 'recurrence element found' if @calendar.service.debug
          if @recurrence
            puts 'setting recurrence' if @calendar.service.debug
            ele.text = @recurrence.to_s
          else
            puts 'no recurrence, adding when' if @calendar.service.debug
            w = xml.root.add_element("gd:when")
            xml.root.delete_element("/entry/gd:recurrence")
            w.attributes["startTime"] = @all_day ? @start.strftime("%Y-%m-%d") : @start.xmlschema
            w.attributes["endTime"] = @all_day ? @end.strftime("%Y-%m-%d") : @end.xmlschema
            set_reminder(w)
          end
        end
      end        
      if not @attendees.empty?
        @attendees.each do |a|
          xml.root.add_element("gd:who", {"email" => a[:email], "valueString" => a[:name], "rel" => "http://schemas.google.com/g/2005#event.attendee"})
        end
      end
      xml.to_s
    end
    
    #Loads the event info from an XML string.
    def load(string)
      @xml = string
      @exists = true
      xml = REXML::Document.new(string)
      @etag = xml.root.attributes['etag']
      xml.root.elements.each(){}.map do |ele|
          case ele.name
             when 'updated'
                @updated = ele.text
             when 'published'
                @published = ele.text
             when 'edited'
                @edited = ele.text
             when 'id'
                @id, @edit_feed = ele.text
             when 'title'
                @title = ele.text
              when 'content'
                @content = ele.text
              when "when"
                @start = Time.parse(ele.attributes['startTime'])
                @end = Time.parse(ele.attributes['endTime'])
                ele.elements.each("gd:reminder") do |r|
                  @reminder = {:minutes => r.attributes['minutes'] ? r.attributes['minutes'] : 0, :hours => r.attributes['hours'] ? r.attributes['hours'] : 0, :days => r.attributes['days'] ? r.attributes['days'] : 0, :method => r.attributes['method'] ? r.attributes['method'] : ''}
                end
              when "where"
                @where = ele.attributes['valueString']
              when "link"
                if ele.attributes['rel'] == 'edit'
                  @edit_feed = ele.attributes['href']
                end
              when "who"
                if ele.attributes['rel'] == "http://schemas.google.com/g/2005#event.attendee"
                n = {}
                ele.attributes.each do |name, value|
                    case name
                      when "email"
                        n[:email] = value
                      when "valueString"
                        n[:name] = value
                    end
                  end                
               @attendees << n
               end
              when "eventStatus"
              case ele.attributes["value"] 
                when "http://schemas.google.com/g/2005#event.confirmed"
                 @status =  :confirmed
                when "http://schemas.google.com/g/2005#event.tentative"
                  @status = :tentative
                when "http://schemas.google.com/g/2005#event.cancelled"
                  @status = :cancelled
              end
            when 'recurrence'
              @recurrence = Recurrence.new(ele.text)
            when "transparency"
               case ele.attributes["value"]
                  when "http://schemas.google.com/g/2005#event.transparent" 
                    @transparency = :free
                  when "http://schemas.google.com/g/2005#event.opaque"
                    @transparency = :busy
                end
            end      
        end
    end
    
    #Reloads the event data from the Google Calendar Service.  Returns true if successful,
    #false otherwise.
    def reload
      t = Event.find(@calendar, @id)
      if t
        if load(t.to_xml)
         return true
        else
         return false
        end
      else
        return false
      end
    end
    
    #Finds the event that matches a query term in the event title or description.
    #  
    #'query' is a string to perform the search on or an event id.
    # 
    #The params hash can contain the following hash values
    #* *scope*: may be :all or :first, indicating whether to return the first record found or an array of all records that match the query.  Default is :all.
    #* *range*: a hash including a :start and :end time to constrain the search by
    #* *max_results*: an integer indicating the number of results to return.  Default is 25.
    #* *sort_order*: either 'ascending' or 'descending'.
    #* *single_events*: either 'true' to return all recurring events as a single entry, or 'false' to return all recurring events as a unique event for each recurrence.
    #* *ctz*: the timezone to return the event times in
    def self.find(calendar, query = '', params = {})
      query_string = ''
      
      begin 
        test = URI.parse(query).scheme
      rescue Exception => e
        test = nil
      end
      
      if test
        puts "id passed, finding event by id" if calendar.service.debug
        puts "id = "+query if calendar.service.debug
        event_id = query.gsub("/events/","/private/full/") #fix provided by groesser3
      
        es = calendar.service.send_get(event_id)
        puts es.inspect if calendar.service.debug
        if es
          entry = REXML::Document.new(es.read_body).root
          puts 'event found' if calendar.service.debug
          Event.define_xml_namespaces(entry)
          event = Event.new(calendar)
          event.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
          return event
        end
        return nil
      end

  
      #parse params hash for values
      range = params[:range] || nil
      max_results = params[:max_results] || nil
      sort_order = params[:sortorder] || nil
      single_events = params[:singleevents] || nil
      timezone = params[:ctz] || nil
      
      #set up query string
      query_string += "q=#{CGI.escape(query)}" if query
      if range
        if not range.is_a? Hash or (range.size > 0 and (not range[:start].is_a? Time or not range[:end].is_a? Time))
          raise "The date range must be a hash including the :start and :end date values as Times"
        else
          date_range = ''
          if range.size > 0
            #Added via patch from Fabio Inguaggiato
            query_string += "&start-min=#{CGI::escape(range[:start].xmlschema)}&start-max=#{CGI::escape(range[:end].xmlschema)}"
          end
        end
      end
      query_string += "&max-results=#{max_results}" if max_results
      query_string += "&sortorder=#{sort_order}" if sort_order
      query_string += "&ctz=#{timezone.gsub(" ", "_")}" if timezone
      query_string += "&singleevents=#{single_events}" if single_events
      if query_string
        events = calendar.service.send_get("http://www.google.com/calendar/feeds/#{calendar.id}/private/full?"+query_string)
        ret = []
        REXML::Document.new(events.read_body).root.elements.each("entry"){}.map do |entry|
          Event.define_xml_namespaces(entry)
          event = Event.new(calendar)
          event.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
          ret << event
        end
      end
      if params[:scope] == :first
        return ret[0]
      else
        return ret
      end
    end
    
    #Returns true if the event exists on the Google Calendar Service.
    def exists?
      return @exists
    end
  
    private 
    @exists = false
    @calendar = nil
    @xml = nil
    @etag = nil
    @recurrence = nil
    @deleted = false
    @edit_feed = ''
    
    def self.define_xml_namespaces(entry)
      entry.attributes["xmlns:gCal"] = "http://schemas.google.com/gCal/2005"
      entry.attributes["xmlns:gd"] = "http://schemas.google.com/g/2005"
      entry.attributes["xmlns:app"] = "http://www.w3.org/2007/app"
      entry.attributes["xmlns"] = "http://www.w3.org/2005/Atom"
      entry.attributes["xmlns:georss"] = "http://www.georss.org/georss"
      entry.attributes["xmlns:gml"] = "http://www.opengis.net/gml"
    end
    
    def set_reminder(ele)
      ele.delete_element("gd:reminder")
      if @reminder
        e = ele.add_element("gd:reminder")
        used = false
        if @reminder[:minutes] 
          e.attributes['minutes'] = @reminder[:minutes] 
          used = true
        elsif @reminder[:hours] and not used
          e.attributes['hours'] = @reminder[:hours]
          used = true
        elsif @reminder[:days] and not used
          e.attributes['days'] = @reminder[:days]
        end
        if @reminder[:method] 
          e.attributes['method'] = @reminder[:method]
        else
          e.attributes['method'] = 'email'
        end
      else
        ele.delete_element("gd:reminder")
      end
    end
  end
end

